test/orange/surf/detail/InvoluteSolver.test.cc

```diff
@@ -628,7 +628,7 @@ TEST(Components, line_angle_param)
     // beta = -pi/3
     {
         real_type u = 0.5;
-        real_type v = std::sin(pi / 3);
+        real_type v = std::sin(pi / real_type{3});
```

Comment:
This (type of) change gives me pause.  The intent of 'real_type' being set to `float` is to switch calculation to using the single floating point processors.  In order for this to happen properly and consistently, with the current PR code, we must never remember to add the explicit casting to real_type of all integer that are use with constants (and then passed to a function); as in this change.

It is likely that we might let in the code the 'obvious' spelling, i.e.: `real_type v = std::sin(pi / 3);` (which currently using the double version of `std::sin`).

The question is whether this (potential) cost is worth the advantage (i.e. if I understand correctly, being able to write `Constant x = y * 2;` rather than `Constant y = y * 2.0` to get the constant calculation in double).

We "can" catch both type of errors during PR review.  It feels that the `Constant` creation is going to be 'rarer' (and more obvious) than the usage inside formula, so maybe we should switch the operation with integer to default to `real_type` rather than `double`.

Reply:
Actually the way I rewrote `pi` at the start of this file, `constexpr real_type pi{constants::pi};`, means this change is unnecessary. I think my initial attempt in that testing file was to reference the constant instead of define a local `real_type` to preserve the same results.

Below is what I wrote under the misapprehension that `pi` was a constant:

---

> This (type of) change gives me pause. The intent of 'real_type' being set to float is to switch calculation to using the single floating point processors. In order for this to happen properly and consistently, with the current PR code, we must never remember to add the explicit casting to real_type of all integer that are use with constants (and then passed to a function); as in this change.

Not here: since the standard library doesn't take Constant as operators, this would be a compiler error without the cast, and that forces us to be specific about whether this is a "setup time"/host calculation or "runtime" calculation.

> It is likely that we might let in the code the 'obvious' spelling, i.e.: `real_type v = std::sin(pi / 3)`; (which currently using the double version of std::sin).

Even if `std::sin` was a function that takes (and returns) a Constant, we can't implicitly cast to a `real_type`.

> so maybe we should switch the operation with integer to default to real_type rather than double

This seems odd too, because `2` is higher precision than `2.0` which is higher precision than `2.0f` (aka possibly `real_type{2}`). I think demoting to runtime precision is the wrong move there.

Since it's ambiguous, what we *could* do is completely eliminate operations with integers, forcing the explicit use of `x = y * Constant{2}`, `x < Constant{0}`, etc.: but I judged that caused more code changes and was a little awkward.

Reply:
> . I think demoting to runtime precision is the wrong move there.

But with `pi` being a real_type`, both `pi / 3` and `pi / real_type{3}` do the same thing (demote to float).

> rewrote pi at the start of this file, `constexpr real_type pi{constants::pi};`

I missed that.  I think it actually make the code much harder to understand/read.  If one must reduce the code length, I would do at most: `constexpr auto pi = constant::pi;`, i.e. I would ***not*** change the type (and If you must change the type, I would change the name too `constexpr real_type real_type_pi {constants::pi};`

Anyway, there is 2 types of usages I had in mind:
```
(a) Constant 3pi = constant::pi * three;
(b) real_type v = overloaded_numerical_function( constant_pi * three );
```
The question is how to spell `three` (or anything else) in both case.  As you pointed out `(a)` is nice with the current PR:
```
 Constant 3pi = constant::pi * 3;
```
but `(b)` fails to compile because there is no implicit conversion.  So one possible spelling with the current PR is
```
real_type v = overloaded_numerical_function( real_type{constant::pi * three} );
```
This looks fine albeit a bit annoying.  Did you consider making the conversion to `real_type` implicit? (it probably would require so `#ifdef` :(  in the `Constant` definition)

Reply:
Since it was a unit test I wanted to minimize changes over the file, especially since evaluating at a different precision level would change the answers.

For (b) I think explicitly casting is the better approach because a constant is "full precision" (we could even make them long doubles or use some `constexpr` extended precision class to get more accurate combined "constant" expressions) but we need to evaluate it in a floating point context: so telling explicitly which level of floating point precision actually improves matters IMHO.

Reply:
> Since it was a unit test I wanted to minimize changes over the file

I think that is counter-productive here.  The tests should follow the same patterns that we would want in the code otherwise they pollute various code investigation (including but not limiting to 'grepping for stuff' but also learn how things should work in Celeritas).

I.e. I think the PR is good as is except for needing to remove any 're-declaration' of constants to not be `Constant`

Reply:
That's a very good point. I'll evaluate the changes...